distsys/resources/relaxedmailboxes.go

```diff
@@ -0,0 +1,267 @@
+)
+
+// RelaxedMailboxesMaker produces a distsys.ArchetypeResourceMaker for a
+// collection of TCP mailboxes. However, relaxed mailboxes don't follow 2PC
+// semantics strictly same as TCP mailboxes. The main difference is that
+// when a critical section successfully sends a message using relaxed
+// mailboxes (res.Write returns with no error), it will be not possible
+// to abort that critical section anymore. Therefore, it's not always safe
+// to use relaxed mailboxes instead of TCP mailboxes. It's only safe to use
+// them when there is at most one network send operation and all other
+// operations are guaranteed to commit successfully.
+func RelaxedMailboxesMaker(addressMappingFn MailboxesAddressMappingFn) distsys.ArchetypeResourceMaker {
+	return IncrementalMapMaker(func(index tla.TLAValue) distsys.ArchetypeResourceMaker {
+		typ, addr := addressMappingFn(index)
+		switch typ {
+		case MailboxesLocal:
+			return relaxedMailboxesLocalMaker(addr)
+		case MailboxesRemote:
+			return relaxedMailboxesRemoteMaker(addr)
+		default:
+			panic(fmt.Errorf("invalid mailbox type %d for address %s: expected local or remote, which are %d or %d", typ, addr, MailboxesLocal, MailboxesRemote))
+		}
+	})
+}
+
+type relaxedMailboxesLocal struct {
+	distsys.ArchetypeResourceLeafMixin
+	listenAddr string
+	msgChannel chan tla.TLAValue
+	listener   net.Listener
+
+	readBacklog     []tla.TLAValue
+	readsInProgress []tla.TLAValue
+
+	done chan struct{}
+}
+
+var _ distsys.ArchetypeResource = &relaxedMailboxesLocal{}
+
+func relaxedMailboxesLocalMaker(listenAddr string) distsys.ArchetypeResourceMaker {
+	return distsys.ArchetypeResourceMakerFn(func() distsys.ArchetypeResource {
+		msgChannel := make(chan tla.TLAValue, mailboxesReceiveChannelSize)
+		listener, err := net.Listen("tcp", listenAddr)
+		if err != nil {
+			panic(fmt.Errorf("could not listen on address %s: %w", listenAddr, err))
+		}
+		log.Printf("started listening on: %s", listenAddr)
+		res := &relaxedMailboxesLocal{
+			listenAddr: listenAddr,
+			msgChannel: msgChannel,
+			listener:   listener,
+			done:       make(chan struct{}),
+		}
+		go res.listen()
+
+		return res
+	})
+}
+
+func (res *relaxedMailboxesLocal) listen() {
+	for {
+		conn, err := res.listener.Accept()
+		if err != nil {
+			select {
+			case <-res.done:
+				return
+			default:
+				panic(fmt.Errorf("error listening on %s: %w", res.listenAddr, err))
+			}
+		}
+		go res.handleConn(conn)
+	}
+}
+
+func (res *relaxedMailboxesLocal) handleConn(conn net.Conn) {
+	defer func() {
+		err := conn.Close()
+		if err != nil {
+			log.Printf("error closing connection: %v", err)
+		}
+	}()
+
+	var err error
+	decoder := gob.NewDecoder(conn)
+	for {
+		if err != nil {
+			select {
+			case <-res.done:
+			default:
+				if err != io.EOF {
+					log.Printf("network error during handleConn, dropping connection: %s", err)
+				}
+			}
+			return
+		}
+		var value tla.TLAValue
+		errCh := make(chan error)
+		go func() {
```

Comment:
I notice this is both here and in the original code, and... I'm not sure why we're spawning a goroutine here.

Wouldn't the Decode call fail fast anyway if we were shutting down, causing us to just end up at line 99 and exit, without needing extra channels + threading?

Reply:
The point is that when a goroutine is blocked on reading from a connection, if another goroutine closes the shared listener, it doesn't cause the first goroutine to get unblocked. The first goroutine still will be blocked.

Reply:
Ahh, I see. So this is to allow the select statement below, which forces an early return (which then closes the connection, unblocking the goroutine)... and this wouldn't happen otherwise, because we can't just close the connection, because we might still need it for finishing a commit (see several previous design discussions).

Thanks for explaining, I did not see that detail.